### PR TITLE
Add support for discovering separately installed exporters using entry points

### PR DESCRIPTION
**Pull request by takluyver**
The group name to advertise additional exporters is `nbconvert.exporter`.

This adds a dependency on my (tiny) [entrypoints](http://entrypoints.readthedocs.org/en/latest/index.html) module.

### CODE REVIEW

**Comment by ellisonbg** [comment]
With the nbextention stuff we decided to not use entry points. @minrk what are your thought on this usage ?

**Comment by minrk** [comment]
Unlike nbextensions, I think entrypoints do make sense for nbconvert, where listing options makes more sense to me. It also allows exporters to pick their names, enabling things like `jupyter nbconvert --to asciidoc`, which I don't think is as relevant to nbextensions.

I view nbconvert exporters as being similar to pygments formatters, which uses entrypoints, whereas nbextensions are more like regular packages.

**Comment by ellisonbg** [comment]
k, great!

On Wed, Feb 3, 2016 at 7:34 AM, Min RK <notifications@github.com> wrote:

> Unlike nbextensions, I think entrypoints do make sense for nbconvert,
> where listing options makes more sense to me. It also allows exporters to
> pick their names, enabling things like jupyter nbconvert --to asciidoc,
> which I don't think is as relevant to nbextensions.
>
> I view nbconvert exporters as being similar to pygments formatters, which
> uses entrypoints, whereas nbextensions are more like regular packages.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/nbconvert/pull/227#issuecomment-179296029>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by Carreau** [comment]
+1

we are slowly eating setuptools marketshare and features !

**Comment by damianavila** [comment]
+1 as well

**Comment by Carreau** [comment]
:-( unfortunately need rebase.

**Comment by takluyver** [comment]
Rebased.

**Comment by takluyver** [comment]
Ooh, we now get a 'Project Jupyter member' badge on our comments. That's new.

**Comment by Carreau** [comment]
> Ooh, we now get a 'Project Jupyter member' badge on our comments. That's new.

That replace the "owner" thing.

Merging, thanks.

**Comment by jezdez** [comment]
FWIW, since `entrypoints` isn't available as a source distribution on PyPI this change breaks using setuptools' `test_require` parameter and its `python setup.py test` command line tool for any Jupyter extension.

When it tries to fulfill the test requirements setuptools looks for suitable releases and is not available as a source distribution but only as a wheel distribution, it'll give up. I'm not sure what the best course of action here is, just wanted to let you know that.

Personally I'd recommend using setuptools' own entrypoints, but uploading a source distribution to PyPI for the entrypoints library would also work. Any opinion on that @takluyver?

**Comment by takluyver** [comment]
There are PEPs in the works for a new sdist format that won't require setup.py. The official position now seems to be that `setup.py` should only ever be invoked through pip, never directly.

**Comment by jezdez** [comment]
@takluyver I'm aware of the PEPs but with all due respect it seems as if you're jumping the gun a bit with only supporting whl files.

FWIW I fully understand and support your design decisions for flit, I've "fixed" way too many setup.py files in my own time and still recover from working on pip for a while. But I'm worried that basing a dependency for Jupyter -- especially as it was mentioned as "tiny" in this PR -- adds unnecessary risk for installation issues.

In other words, I'm not sure if Jupyter should be trying solve packaging problems on its own but wait for the upstream tools like pip adopt the PEPs. Until then I'd recommend reducing complexity and go with what currently works, e.g. via setuptools' entrypoints or creating a sdist for the entrypoints library.

**Comment by takluyver** [comment]
I think the PEPs can formalise a solution that people are already experimenting with, but if we all sit back and wait for the PEPs, there won't be enough pressure to make those PEPs happen.

In the meantime, setuptools should AIUI see a package once it's been installed from a wheel, so it should be possible to work around it by running `pip install .` before testing.

**Comment by jezdez** [comment]
Totally agreed about PEP and building pressure to finally getting rid of the hacks on top of hacks!

That's a great work-around, I hadn't thought about that, it was even enough to add `entrypoints` as a requirement to the tox config (that the project in particular has as well and which calls `python setup.py test`), so setuptools won't try to fulfill that dependency on its own. Thanks and cheers!

**Comment by takluyver** [comment]
Excellent, glad it's working out :-)
